### DISCUSSION TITLE
Generating paths from package names

### DISCUSSION THREAD

**Discussion by steveukx** [jpmorganchase/modular#2104] [General] [answered]
In `modular add PKG` we create a new package in the monorepo with a `name` field in the `package.json` derived from the `PKG` input, and in a directory path also derived from the `PKG` input.

Use cases to be resolved:

1. When a user wants to add an unscoped package to the packages root
1. When a user wants to add an unscoped package to a nested path 
1. When a user wants to add a scoped package
1. When a user wants to add a scoped package to a nested path
1. When a user wants to add a package (scoped or otherwise) to a path other than `./packages`

**Comment by steveukx** [comment]
```
# adds "PKG" in ./packages/PKG 
modular add PKG

# adds "path-to-PKG" in ./packages/path/to/PKG
modular add path/to/PKG

# adds "@scope/PKG" in ./packages/PKG
modular add @scope/PKG

# adds "@scope/path-to-PKG" in ./packages/path/to/PKG
modular add path/to/@scope/PKG

# adds "PKG" in ./foo/PKG
modular add PKG --path ./foo

# adds "slash-prefixed-PKG" in ./foo/slash/prefixed/PKG
modular add slash/prefixed/PKG --path ./foo
```

Assuming `--path` contains a relative directory (would need to validate doesn't use `..` blocks to jump outside of repo root) that can be used in place of the default `./packages`.

Caveat of using a `--path`, modular should check that the workspace would find the newly created package with the `workspaces` property in root `package.json`

**Comment by cristiano-belloni** [comment]
Simplest proposition:

```
# adds "@scope/PKG" in ./packages/scope/PKG
modular add @scope/PKG

# adds "PKG" in ./packages/PKG 
modular add PKG

# adds "PKG" in ./foo/PKG
modular add PKG --path ./foo

# adds "@scope/PKG" in ./foo/scope/PKG
modular add @scope/PKG --path ./foo

# Throws if PKG is an invalid package name
modular add path/to/@scope/PKG
[modular] Error: provided workspace name "path/to/@scope/PKG" is invalid.

```

**Comment by cristiano-belloni** [comment]
Simplest proposition with use-cases:

1. When a user wants to add an unscoped package to the packages root
```
# adds "PKG" in ./packages/PKG
modular add PKG
```
2. When a user wants to add an unscoped package to a nested path
```
# adds "PKG" in packages/my/nested/path/PKG
modular add PKG --path packages/my/nested/path
```
3. When a user wants to add a scoped package
```
# adds "@scope/PKG" in packages/scope/PKG
modular add @scope/PKG
```
4. When a user wants to add a scoped package to a nested path
```
# adds "@scope/PKG" in packages/my/nested/path/scope/PKG
modular add @scope/PKG --path packages/my/nested/path
```
5. When a user wants to add a package (scoped or otherwise) to a path other than ./packages
```
# adds "PKG" in my/other/path/PKG
modular add PKG --path /my/other/path

# adds "@scope/PKG" in my/other/path/scope/PKG
modular add @scope/PKG --path my/other/path
```

Validity checks (from @steveukx 's comment):

- Assuming `--path` contains a relative directory (would need to validate doesn't use `..` blocks to jump outside of repo root) that can be used in place of the default ./packages.

- Caveat of using a `--path`, modular should check that the workspace would find the newly created package with the workspaces property in root package.json

**Comment by cristiano-belloni** [comment, marked as answer]
Revised proposition with use-cases:

1. When a user wants to add an unscoped package to the packages root
```
# adds "pkg" in ./packages/pkg
modular add PKG
```
2. When a user wants to add an unscoped package to a nested path
```
# adds "pkg" in packages/my/nested/path/pkg
modular add pkg --path packages/my/nested/path
```
3. When a user wants to add a scoped package
```
# adds "@scope/pkg" in packages/pkg
modular add @scope/pkg
```
4. When a user wants to add a scoped package to a nested path
```
# adds "@scope/pkg" in packages/my/nested/path/scope/pkg
modular add @scope/pkg --path packages/my/nested/path/scope
```
5. When a user wants to add a package (scoped or otherwise) to a path other than ./packages
```
# adds "pkg" in my/other/path/pkg
modular add pkg --path my/other/path

# adds "@scope/pkg" in my/other/path/pkg
modular add @scope/PKG --path my/other/path
```

Validity checks (from @steveukx 's comment):

- Assuming `--path` contains a relative directory (would need to validate doesn't use `..` blocks to jump outside of repo root) that can be used in place of the default ./packages.

- Caveat of using a `--path`, modular should check that the workspace would find the newly created package with the workspaces property in root package.json

**Reply by benpryke** [reply]
If a user runs `modular add pkg` when the current working directory is `packages/path`, is the `pkg` added to `packages/pkg` or `packages/path/pkg`?

**Reply by cristiano-belloni** [reply]
> If a user runs `modular add pkg` when the current working directory is `packages/path`, is the `pkg` added to `packages/pkg` or `packages/path/pkg`?

In that case modular fails before running the specified command (I think that the preliminary check fails)
(let's open another discussion if needed as this behaviour is not strictly `modular add`-related)

**Reply by sgb-io** [reply]
In example 5, you have two examples, one using `/my/other/path`, the other using `my/other/path`. Is that intentional?

**Reply by cristiano-belloni** [reply]
> In example 5, you have two examples, one using `/my/other/path`, the other using `my/other/path`. Is that intentional?

typo, corrected
